apps/api-documenter/src/markdown/MarkdownEmitter.ts

```diff
@@ -141,9 +141,21 @@ export class MarkdownEmitter {
         const trimmedParagraph: DocParagraph = DocNodeTransforms.trimSpacesInParagraph(docParagraph);
         if (context.insideTable) {
           if (docNodeSiblings) {
-            writer.write('<p>');
+            // This deferred write is necessary to avoid writing empty paragraph tags (i.e. `<p></p>`). At the
+            // time this code runs, we do not know whether the `writeNodes` call below will actually write
+            // anything. Thus, we want to only write a `<p>` tag (as well as eventually a corresponding
+            // `</p>` tag) if something else ends up being subsequently written.
+            writer.deferredWrite('<p>');
             this.writeNodes(trimmedParagraph.nodes, context);
-            writer.write('</p>');
+
+            if (writer.hasDeferred) {
+              // We know nothing has been written after the `deferredWrite`, so clear any deferred messages and
+              // don't write a closing `</p>` tag.
+              writer.clearDeferred();
+            } else {
+              // We know the starting `<p>` tag has been written, so write a closing `</p>` tag.
+              writer.write('</p>');
+            }
```

Comment:
If I understand correctly, the idea of this algorithm is to create a nesting container, but if the contents of the container turn out to be empty, then we backtrack and revert the container.  Theoretically we might write `<table>`, `<tr>`, `<td>` before the recursive traversal knows whether `CONTENT` is empty or not:

```html
<table>
  <tr>
    <td>CONTENT</td>
  <tr>
</table>
```

A couple observations:
- The word "deferred" focuses on *when* the items get written, whereas the key concept is *whether* they get written.  Maybe "tentative" is a better term?
- It seems very likely that such calls can be nested, for example something like `> > > triple nested blockquote`

Maybe the API should be like:

```ts
function f(writer) {
  const tentativeScope2 = writer.beginTentativeWrite('<b>');
  writer.write('CONTENT');
  if (tentativeScope2.end()) {
    writer.write('</b>');
  }
}

const tentativeScope = writer.beginTentativeWrite('<p>');
f(writer);
if (tentativeScope.end()) {
  writer.write('</p>');
}
```

...or if you like dealing with closures:

```ts
function f(writer) {
  writer.writeTentative('<b>', '</b>', () => {
    writer.write('CONTENT');
  });  
}

writer.writeTentative('<p>', '</p>', () => {
  f(writer);
});
```

(These are just suggestions -- if it is a lot of work to implement, I don't feel too strongly about it.)

Reply:
Good ideas - I updated the API to the `writeTentative` closure-version - LMK what you think.